res/res_stir_shaken/verification.c

```diff
@@ -0,0 +1,1028 @@
+	[AST_STIR_SHAKEN_VS_INVALID_GRANT] = "invalid_grant",
+};
+
+const char *ast_stir_shaken_vs_response_code_to_str(
+	enum ast_stir_shaken_vs_response_code vs_rc)
+{
+	return ARRAY_IN_BOUNDS(vs_rc, vs_rc_map) ?
+		vs_rc_map[vs_rc] : NULL;
+}
+
+static void cleanup_cert_from_astdb_and_fs(
+	struct ast_stir_shaken_vs_ctx *ctx)
+{
+	if (ast_db_exists(ctx->hash_family, "path") || ast_db_exists(ctx->hash_family, "expiration")) {
+		ast_db_deltree(ctx->hash_family, NULL);
+	}
+
+	if (ast_db_exists(ctx->url_family, ctx->public_url)) {
+		ast_db_del(ctx->url_family, ctx->public_url);
+	}
+
+	/* Remove the actual file from the system */
+	remove(ctx->filename);
+}
+
+static int add_cert_expiration_to_astdb(struct ast_stir_shaken_vs_ctx *cert,
+	const char *cache_control_header, const char *expires_header)
+{
+	RAII_VAR(struct ss_vs_cfg *, cfg, ss_get_vs_cfg(), ao2_cleanup);
+
+	char time_buf[32];
+	time_t current_time = time(NULL);
+	time_t max_age_hdr = 0;
+	time_t expires_hdr = 0;
+	ASN1_TIME *notAfter = NULL;
+	time_t cert_expires = 0;
+	time_t config_expires = 0;
+	time_t expires = 0;
+	int rc = 0;
+
+	config_expires = current_time + cfg->max_cache_entry_age;
+
+	if (!ast_strlen_zero(cache_control_header)) {
+		char *str_max_age;
+
+		str_max_age = strstr(cache_control_header, "s-maxage");
+		if (!str_max_age) {
+			str_max_age = strstr(cache_control_header, "max-age");
+		}
+
+		if (str_max_age) {
+			unsigned int m;
+			char *equal = strchr(str_max_age, '=');
+			if (equal && !ast_str_to_uint(equal + 1, &m)) {
+				max_age_hdr = current_time + m;
+			}
+		}
+	}
+
+	if (!ast_strlen_zero(expires_header)) {
+		struct ast_tm expires_time;
+
+		ast_strptime(expires_header, "%a, %d %b %Y %T %z", &expires_time);
+		expires_time.tm_isdst = -1;
+		expires_hdr = ast_mktime(&expires_time, "GMT").tv_sec;
+	}
+
+	notAfter = X509_get_notAfter(cert->xcert);
+	cert_expires = ast_crypto_asn_time_as_time_t(notAfter);
+
+	/*
+	 * We got all the possible values of expires so
+	 * let's find the lowest non-zero value.
+	 */
+
+	/* The default */
+	expires = config_expires;
+
+	if (max_age_hdr && max_age_hdr < expires) {
+		expires = max_age_hdr;
+	}
+
+	if (expires_hdr && expires_hdr < expires) {
+		expires = expires_hdr;
+	}
+
+	if (cert_expires && cert_expires < expires) {
+		expires = cert_expires;
+	}
+
+	snprintf(time_buf, sizeof(time_buf), "%ld", expires);
+
+	rc = ast_db_put(cert->hash_family, "expiration", time_buf);
+	if (rc == 0) {
+		strcpy(cert->expiration, time_buf); /* safe */
+	}
+
+	return rc;
+}
+
+static int add_cert_key_to_astdb(struct ast_stir_shaken_vs_ctx *cert,
+	const char *cache_control_hdr, const char *expires_hdr)
+{
+	int rc = 0;
+
+	rc = ast_db_put(cert->url_family, cert->public_url, cert->hash);
+	if (rc) {
+		return rc;
+	}
+	rc = ast_db_put(cert->hash_family, "path", cert->filename);
+	if (rc) {
+		ast_db_del(cert->url_family, cert->public_url);
+		return rc;
+	}
+
+	rc = add_cert_expiration_to_astdb(cert, cache_control_hdr, expires_hdr);
```

Comment:
Cleanup here as well if this fails (ast_db_del the previous two)

Reply:
done